File: frontend/js/views/fifty_thirty_twenty.js
// Elizabeth Warren's 50/30/20 rule: split after-tax income into
// 50% needs, 30% wants, 20% savings + debt principal. Each input
// row classifies into one bucket. Compute reports per-bucket actual
// vs ideal + delta + status.

import { api } from '../api.js';
import { esc } from '../util.js';
import { t } from '../i18n.js';

const STATE = {
    net_monthly_income_usd: 5000,
    rows: [
        { name: 'Rent / mortgage',  bucket: 'needs',   amount_usd: 1800 },
        { name: 'Groceries',        bucket: 'needs',   amount_usd: 500 },
        { name: 'Utilities',        bucket: 'needs',   amount_usd: 180 },
        { name: 'Insurance',        bucket: 'needs',   amount_usd: 200 },
        { name: 'Dining out',       bucket: 'wants',   amount_usd: 400 },
        { name: 'Streaming',        bucket: 'wants',   amount_usd: 60 },
        { name: 'Hobbies',          bucket: 'wants',   amount_usd: 300 },
        { name: '401k',             bucket: 'savings', amount_usd: 800 },
        { name: 'Roth IRA',         bucket: 'savings', amount_usd: 200 },
    ],
};

export async function renderFiftyThirtyTwenty(mount, _state) {
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.fifty_thirty_twenty.title">// 50/30/20 BUDGET RULE</span></h1>
        <p class="muted small" data-i18n-html="view.fifty_thirty_twenty.intro">
            Elizabeth Warren's 50/30/20 rule (<em>All Your Worth</em>, 2005). Split after-tax
            (net) income: <strong>50% needs</strong> (rent, groceries, utilities, insurance,
            minimum debt), <strong>30% wants</strong> (dining out, streaming, hobbies),
            <strong>20% savings</strong> (emergency fund, retirement, extra debt principal).
            For needs/wants <em>under</em> ideal is on-track; for savings <em>at-or-above</em>
            ideal is on-track.
        </p>
        <div class="chart-panel">
            <div style="display:flex;gap:8px;align-items:end;margin-bottom:12px">
                <label style="flex:0 0 240px">
                    <span class="muted small" data-i18n="view.fifty_thirty_twenty.field.income">Net monthly income $</span>
                    <input type="number" id="fty-income" step="100" min="0" value="${STATE.net_monthly_income_usd}" style="width:100%">
                </label>
                <button class="btn btn-sm primary" id="fty-run" data-shortcut="r" data-i18n="view.fifty_thirty_twenty.btn.run">⚡ Compute Split</button>
            </div>
            <h2>${esc(t('view.fifty_thirty_twenty.h2.rows'))}</h2>
            <table class="trades" id="fty-table">
                <thead><tr>
                    <th data-i18n="view.fifty_thirty_twenty.th.name">Item</th>
                    <th data-i18n="view.fifty_thirty_twenty.th.bucket">Bucket</th>
                    <th data-i18n="view.fifty_thirty_twenty.th.amount">Amount $</th>
                    <th></th>
                </tr></thead>
                <tbody id="fty-body"></tbody>
            </table>
            <button class="btn btn-sm" id="fty-add" data-i18n="view.fifty_thirty_twenty.btn.add">＋ Add row</button>
            <div id="fty-result"></div>
        </div>
    `;
    drawRows(mount);
    mount.querySelector('#fty-income').addEventListener('input', e => {
        STATE.net_monthly_income_usd = parseFloat(e.target.value) || 0;
    });
    mount.querySelector('#fty-add').addEventListener('click', () => {
        STATE.rows.push({ name: 'New row', bucket: 'wants', amount_usd: 0 });
        drawRows(mount);
    });
    mount.querySelector('#fty-run').addEventListener('click', () => runCompute(mount));
    await runCompute(mount);
}

function drawRows(mount) {
    const body = mount.querySelector('#fty-body');
    body.innerHTML = STATE.rows.map((r, i) => `
        <tr>
            <td><input type="text" data-k="name" data-i="${i}" value="${esc(r.name)}" style="width:100%"></td>
            <td>
                <select data-k="bucket" data-i="${i}">
                    <option value="needs"   ${r.bucket === 'needs'   ? 'selected' : ''}>Needs (50%)</option>
                    <option value="wants"   ${r.bucket === 'wants'   ? 'selected' : ''}>Wants (30%)</option>
                    <option value="savings" ${r.bucket === 'savings' ? 'selected' : ''}>Savings (20%)</option>
                </select>
            </td>
            <td><input type="number" step="25" min="0" data-k="amount_usd" data-i="${i}" value="${r.amount_usd}" style="width:100%"></td>
            <td><button class="btn btn-xs" data-del="${i}">✕</button></td>
        </tr>
    `).join('');
    body.querySelectorAll('input,select').forEach(inp => {
        const ev = inp.tagName === 'SELECT' ? 'change' : 'input';
        inp.addEventListener(ev, () => {
            const i = parseInt(inp.dataset.i, 10);
            const k = inp.dataset.k;
            STATE.rows[i][k] = k === 'amount_usd' ? (parseFloat(inp.value) || 0) : inp.value;
        });
    });
    body.querySelectorAll('button[data-del]').forEach(btn => {
        btn.addEventListener('click', () => {
            STATE.rows.splice(parseInt(btn.dataset.del, 10), 1);
            drawRows(mount);
        });
    });
}

async function runCompute(mount) {
    const result = mount.querySelector('#fty-result');
    result.innerHTML = `<p class="muted">${esc(t('view.fifty_thirty_twenty.status.computing'))}</p>`;
    try {
        const r = await api('/fifty-thirty-twenty/compute', { method: 'POST', body: JSON.stringify(STATE) });
        const overallCls = r.overall_status === 'on-track' ? 'pos' : 'neg';
        const bucketRow = b => `
            <tr>
                <td><strong style="text-transform:uppercase">${esc(t('view.fifty_thirty_twenty.bucket.' + b.bucket) || b.bucket)}</strong></td>
                <td>${b.ideal_pct.toFixed(0)}%</td>
                <td>$${b.ideal_usd.toFixed(0)}</td>
                <td>$${b.actual_usd.toFixed(0)}</td>
                <td>${b.actual_pct.toFixed(1)}%</td>
                <td class="${b.delta_usd > 0 ? 'neg' : 'pos'}">${b.delta_usd >= 0 ? '+' : ''}$${b.delta_usd.toFixed(0)}</td>
                <td class="${b.status === 'on-track' ? 'pos' : 'neg'}" style="text-transform:uppercase"><strong>${esc(t('view.fifty_thirty_twenty.status.' + b.status.replace('-', '_')) || b.status)}</strong></td>
            </tr>
        `;
        result.innerHTML = `
            <div style="margin-top:1rem">
                <div class="muted small">${esc(t('view.fifty_thirty_twenty.field.overall'))}</div>
                <strong class="${overallCls}" style="font-size:1.4em;text-transform:uppercase">${esc(t('view.fifty_thirty_twenty.status.' + r.overall_status.replace('-', '_')) || r.overall_status)}</strong>
                <span class="muted" style="margin-left:1rem">${esc(t('view.fifty_thirty_twenty.field.unallocated'))}: <strong>$${r.unallocated_usd.toFixed(0)}</strong></span>
            </div>
            <h2 style="margin-top:1rem">${esc(t('view.fifty_thirty_twenty.h2.buckets'))}</h2>
            <table class="trades">
                <thead><tr>
                    <th data-i18n="view.fifty_thirty_twenty.th.bucket_h">Bucket</th>
                    <th data-i18n="view.fifty_thirty_twenty.th.ideal_pct">Ideal %</th>
                    <th data-i18n="view.fifty_thirty_twenty.th.ideal_usd">Ideal $</th>
                    <th data-i18n="view.fifty_thirty_twenty.th.actual_usd">Actual $</th>
                    <th data-i18n="view.fifty_thirty_twenty.th.actual_pct">Actual %</th>
                    <th data-i18n="view.fifty_thirty_twenty.th.delta">Delta</th>
                    <th data-i18n="view.fifty_thirty_twenty.th.status">Status</th>
                </tr></thead>
                <tbody>
                    ${bucketRow(r.needs)}
                    ${bucketRow(r.wants)}
                    ${bucketRow(r.savings)}
                </tbody>
            </table>
        `;
    } catch (e) {
        result.innerHTML = `<p class="neg">${esc(String(e))}</p>`;
    }
}

File: crates/traderview-db/src/fifty_thirty_twenty.rs
//! 50/30/20 budget rule (Elizabeth Warren, "All Your Worth", 2005).
//!
//! Split after-tax (net) income into three buckets:
//!   - 50% NEEDS    — rent, groceries, utilities, insurance, minimum
//!                    debt payments, mandatory expenses
//!   - 30% WANTS    — dining out, streaming, hobbies, vacations,
//!                    upgrades to needs (organic groceries, gym)
//!   - 20% SAVINGS  — emergency fund, retirement, brokerage, EXTRA
//!                    debt payment beyond minimum
//!
//! Each input row carries a `bucket` ∈ {needs, wants, savings} and
//! an amount. Compute returns per-bucket actual + ideal + delta + %.
//!
//! Pure compute.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct ExpenseRow {
    pub name: String,
    pub bucket: String,
    pub amount_usd: f64,
}

#[derive(Debug, Clone, Deserialize)]
pub struct FiftyThirtyTwentyInput {
    pub net_monthly_income_usd: f64,
    #[serde(default)]
    pub rows: Vec<ExpenseRow>,
}

#[derive(Debug, Clone, Serialize)]
pub struct BucketResult {
    pub bucket: &'static str,
    pub ideal_pct: f64,
    pub ideal_usd: f64,
    pub actual_usd: f64,
    pub actual_pct: f64,
    pub delta_usd: f64,
    pub status: &'static str,
}

#[derive(Debug, Clone, Serialize)]
pub struct FiftyThirtyTwentyReport {
    pub net_monthly_income_usd: f64,
    pub needs: BucketResult,
    pub wants: BucketResult,
    pub savings: BucketResult,
    pub total_allocated_usd: f64,
    pub unallocated_usd: f64,
    pub overall_status: String,
}

// ─── Pure compute ─────────────────────────────────────────────────────────

pub fn sum_for(rows: &[ExpenseRow], bucket: &str) -> f64 {
    rows.iter()
        .filter(|r| r.bucket == bucket)
        .map(|r| r.amount_usd)
        .sum()
}

pub fn evaluate_bucket(
    label: &'static str,
    ideal_pct: f64,
    net_income: f64,
    actual: f64,
) -> BucketResult {
    let ideal_usd = net_income * ideal_pct / 100.0;
    let actual_pct = if net_income > 0.0 {
        actual / net_income * 100.0
    } else {
        0.0
    };
    let delta = actual - ideal_usd;
    let status: &'static str = if label == "savings" {
        // For savings, MORE is better — under is bad.
        if actual >= ideal_usd { "on-track" } else { "under" }
    } else {
        // For needs/wants, under is better.
        if actual <= ideal_usd { "on-track" } else { "over" }
    };
    BucketResult {
        bucket: label,
        ideal_pct,
        ideal_usd,
        actual_usd: actual,
        actual_pct,
        delta_usd: delta,
        status,
    }
}

pub fn compute(input: &FiftyThirtyTwentyInput) -> FiftyThirtyTwentyReport {
    let needs_actual = sum_for(&input.rows, "needs");
    let wants_actual = sum_for(&input.rows, "wants");
    let savings_actual = sum_for(&input.rows, "savings");
    let needs = evaluate_bucket("needs", 50.0, input.net_monthly_income_usd, needs_actual);
    let wants = evaluate_bucket("wants", 30.0, input.net_monthly_income_usd, wants_actual);
    let savings =
        evaluate_bucket("savings", 20.0, input.net_monthly_income_usd, savings_actual);
    let total_alloc = needs_actual + wants_actual + savings_actual;
    let unalloc = input.net_monthly_income_usd - total_alloc;
    let bucket_statuses = [needs.status, wants.status, savings.status];
    let overall = if bucket_statuses.iter().all(|s| *s == "on-track") {
        "on-track"
    } else {
        "off-target"
    }
    .to_string();
    FiftyThirtyTwentyReport {
        net_monthly_income_usd: input.net_monthly_income_usd,
        needs,
        wants,
        savings,
        total_allocated_usd: total_alloc,
        unallocated_usd: unalloc,
        overall_status: overall,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn r(name: &str, bucket: &str, amt: f64) -> ExpenseRow {
        ExpenseRow { name: name.into(), bucket: bucket.into(), amount_usd: amt }
    }

    #[test]
    fn sum_for_basic() {
        let rs = vec![r("a", "needs", 100.0), r("b", "needs", 50.0), r("c", "wants", 30.0)];
        assert_eq!(sum_for(&rs, "needs"), 150.0);
        assert_eq!(sum_for(&rs, "wants"), 30.0);
        assert_eq!(sum_for(&rs, "savings"), 0.0);
    }

    #[test]
    fn evaluate_bucket_needs_on_track_when_under_ideal() {
        let b = evaluate_bucket("needs", 50.0, 5000.0, 2400.0);
        assert_eq!(b.ideal_usd, 2500.0);
        assert_eq!(b.actual_pct, 48.0);
        assert_eq!(b.delta_usd, -100.0);
        assert_eq!(b.status, "on-track");
    }

    #[test]
    fn evaluate_bucket_needs_over_when_above_ideal() {
        let b = evaluate_bucket("needs", 50.0, 5000.0, 2700.0);
        assert_eq!(b.delta_usd, 200.0);
        assert_eq!(b.status, "over");
    }

    #[test]
    fn evaluate_bucket_savings_under_when_below_ideal() {
        let b = evaluate_bucket("savings", 20.0, 5000.0, 800.0);
        assert_eq!(b.ideal_usd, 1000.0);
        assert_eq!(b.status, "under");
    }

    #[test]
    fn evaluate_bucket_savings_on_track_when_at_or_above() {
        let b = evaluate_bucket("savings", 20.0, 5000.0, 1500.0);
        assert_eq!(b.status, "on-track");
    }

    #[test]
    fn evaluate_bucket_zero_income_zero_pct() {
        let b = evaluate_bucket("needs", 50.0, 0.0, 100.0);
        assert_eq!(b.actual_pct, 0.0);
    }

    #[test]
    fn compute_perfectly_balanced() {
        let r = compute(&FiftyThirtyTwentyInput {
            net_monthly_income_usd: 5000.0,
            rows: vec![
                r("rent",     "needs",   2500.0),
                r("fun",      "wants",   1500.0),
                r("401k",     "savings", 1000.0),
            ],
        });
        assert_eq!(r.needs.delta_usd, 0.0);
        assert_eq!(r.wants.delta_usd, 0.0);
        assert_eq!(r.savings.delta_usd, 0.0);
        assert_eq!(r.overall_status, "on-track");
        assert_eq!(r.unallocated_usd, 0.0);
    }

    #[test]
    fn compute_overspending_wants() {
        let r = compute(&FiftyThirtyTwentyInput {
            net_monthly_income_usd: 5000.0,
            rows: vec![
                r("rent",     "needs",   2400.0),
                r("dining",   "wants",   2000.0),  // 40%, should be 30%
                r("401k",     "savings", 600.0),
            ],
        });
        assert_eq!(r.wants.status, "over");
        assert_eq!(r.savings.status, "under");
        assert_eq!(r.overall_status, "off-target");
    }

    #[test]
    fn compute_unallocated_residual() {
        let r = compute(&FiftyThirtyTwentyInput {
            net_monthly_income_usd: 5000.0,
            rows: vec![r("rent", "needs", 1000.0)],
        });
        assert_eq!(r.unallocated_usd, 4000.0);
    }

    #[test]
    fn compute_all_savings_high_status() {
        let r = compute(&FiftyThirtyTwentyInput {
            net_monthly_income_usd: 5000.0,
            rows: vec![
                r("rent", "needs", 2400.0),
                r("fun",  "wants", 1400.0),
                r("401k", "savings", 1500.0),
            ],
        });
        assert_eq!(r.needs.status, "on-track");
        assert_eq!(r.wants.status, "on-track");
        assert_eq!(r.savings.status, "on-track");
        assert_eq!(r.overall_status, "on-track");
    }

    #[test]
    fn compute_unknown_bucket_ignored() {
        let r = compute(&FiftyThirtyTwentyInput {
            net_monthly_income_usd: 5000.0,
            rows: vec![
                r("rent",     "needs",   2500.0),
                r("fun",      "wants",   1500.0),
                r("401k",     "savings", 1000.0),
                r("bogus",    "bogus",   999.0),  // ignored everywhere
            ],
        });
        // None of the bucket sums should include the bogus row.
        assert_eq!(r.needs.actual_usd, 2500.0);
        assert_eq!(r.wants.actual_usd, 1500.0);
        assert_eq!(r.savings.actual_usd, 1000.0);
        // total allocated only counts the three valid rows, not the bogus.
        assert_eq!(r.total_allocated_usd, 5000.0);
    }
}
